core/crypto/nike/csidh/csidh.go

```diff
@@ -0,0 +1,263 @@
+// csidh.go - Adapts csidh module to our NIKE interface.
+// Copyright (C) 2022  David Stainton.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package csidh
+
+import (
+	"encoding/base64"
+	"errors"
+
+	"github.com/henrydcase/nobs/dh/csidh"
+	"github.com/katzenpost/katzenpost/core/crypto/nike"
+	"github.com/katzenpost/katzenpost/core/crypto/rand"
+)
+
+// CSIDHScheme is the nobs CSIDH-512 NIKE.
+var CSIDHScheme *CsidhNike
+
+var _ nike.PrivateKey = (*PrivateKey)(nil)
+var _ nike.PublicKey = (*PublicKey)(nil)
+var _ nike.Scheme = (*CsidhNike)(nil)
+
+type CsidhNike struct{}
+
+func (e *CsidhNike) Name() string {
+	return "CSIDH-512-nobs"
+}
+
+func (e *CsidhNike) PublicKeySize() int {
+	return csidh.PublicKeySize
+}
+
+func (e *CsidhNike) PrivateKeySize() int {
+	return csidh.PrivateKeySize
+}
+
+func (e *CsidhNike) GenerateKeyPair() (nike.PublicKey, nike.PrivateKey, error) {
+	privateKey := new(csidh.PrivateKey)
+	err := csidh.GeneratePrivateKey(privateKey, rand.Reader)
+	if err != nil {
+		return nil, nil, err
+	}
+	publicKey := new(csidh.PublicKey)
```

Comment:
looks like this could use `DerivePublicKey()` to prevent code duplication?

Reply:
yeup DerivePublicKey fits here, thanks